File: 04 Root Finding/newton.f90
! newton.f90  -  find a root by Newton's method
! compile with: gfortran -O3 -fdefault-real-8 newton.f90

program newton
implicit none

real x
integer i

! initial guess
x = 1.0

! Newton's iteration
do i = 1,8
	x = x - f(x)/df(x)
	write (*,*) x, f(x)
end do

contains

! function to find a root of...
pure function f(x); intent(in) x
	real f, x
	f = cos(x)
end function

! derivative of a function to find a root of
pure function df(x); intent(in) x
	real df, x
	df = -sin(x)
end function

end program

File: 04 Root Finding/newton.py
#!/usr/bin/env python
# find a root of a function using Newton's method

#######################################################################

from math import *

#######################################################################

# find root of a function f(x) = 0
def f(x):
	return cos(x)

def df(x):
	return -sin(x)

# initial guess
x = 1.0

# Newton's method update
for i in range(16):
	x -= f(x)/df(x)
	print(i,x,f(x))
